tools/plugins/com.liferay.ide.project.ui/src/com/liferay/ide/project/ui/upgrade/LiferayUpgradeDescriptorTableViewCustomPart.java

```diff
@@ -0,0 +1,624 @@
+import org.eclipse.sapphire.Value;
+import org.eclipse.sapphire.modeling.Path;
+import org.eclipse.sapphire.modeling.Status;
+import org.eclipse.sapphire.platform.PathBridge;
+import org.eclipse.sapphire.platform.StatusBridge;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.graphics.Color;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.ui.PlatformUI;
+import org.jdom.DocType;
+import org.jdom.Document;
+import org.jdom.JDOMException;
+import org.jdom.input.SAXBuilder;
+import org.jdom.output.XMLOutputter;
+
+/**
+ * @author Simon Jiang
+ */
+
+public class LiferayUpgradeDescriptorTableViewCustomPart extends AbstractLiferayTableViewCustomPart
+{
+
+    private class LiferayDescriptorUpgradeCompre extends LiferayUpgradeCompare
+    {
+
+        private final IPath soruceFile;
+        private final IPath targetFile;
+
+        public LiferayDescriptorUpgradeCompre( final IPath soruceFile, final IPath targetFile, String fileName )
+        {
+            super( PlatformUI.getWorkbench().getActiveWorkbenchWindow(), fileName );
+            this.soruceFile = soruceFile;
+            this.targetFile = targetFile;
+        }
+
+        @Override
+        protected File getSourceFile()
+        {
+            return soruceFile.toFile();
+        }
+
+        @Override
+        protected File getTargetFile()
+        {
+            return targetFile.toFile();
+        }
+    }
+
+    private class LiferayDescriptorUpgradeElement extends TableViewerElement
+    {
+
+        public final String location;
+        public final String descriptorName;
+
+        public LiferayDescriptorUpgradeElement( String name, String context, String location, String descriptorName )
+        {
+            super( name, context );
+            this.location = location;
+            this.descriptorName = descriptorName;
+        }
+    }
+
+    private class NIOSearchFilesVisitor extends SimpleFileVisitor<java.nio.file.Path>
+    {
+
+        File resources;
+
+        String searchFileName = null;
+
+        NIOSearchFilesVisitor( String searchFileName )
+        {
+            this.searchFileName = searchFileName;
+        }
+
+        public File getSearchFile()
+        {
+            return this.resources;
+        }
+
+        @Override
+        public FileVisitResult visitFile( java.nio.file.Path path, BasicFileAttributes attrs ) throws IOException
+        {
+            if( path.endsWith( searchFileName ) )
+            {
+                resources = path.toFile();
+
+                return FileVisitResult.TERMINATE;
+            }
+            return FileVisitResult.CONTINUE;
+        }
+    }
+
+    private final static String[][] DESCRIPTORS_AND_IMAGES = { { "liferay-portlet.xml", "/icons/e16/portlet.png" },
+        { "liferay-display.xml", "/icons/e16/liferay_display_xml.png" },
+        { "service.xml", "/icons/e16/service_xml.png" }, { "liferay-hook.xml", "/icons/e16/hook.png" },
```

Comment:
service_xml.png is the `new` version of the file.  We shouldn't use the `new` one.